bras/bras.1
.\"
.\" This file is part of bras, a program similar to the (in)famous
.\" `make'-utitlity, written in Tcl.
.\"
.\" Copyright (C) 1996-2000 Harald Kirsch
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH BRAS 1  "|VERDATE|" "Kir" "bras user's manual"

.SH NAME
bras -\ utility for rule base command execution

.SH SYNOPSIS
bras [-f brasfile] [-d] [-h] [-n] [-p] [-s] [-v] [-ve] [var=value] [--] [target ...]

.SH OPTIONS
.IP "-f brasfile"
specifies the rule-file to use.
Defaults are 
.I brasfile
and
.I Brasfile
in that order.

.IP -d
show reasoning in addition to normal processing. This option requests a
detailed description of the reasoning performed by bras and is useful
for debugging. (Up to version 0.6.0, -d suppressed the execution of any
commands. Since then the reasoning process was changed in a way that
non-execution of commands changes the results. To return exact
results, -d can no longer suppress execution of commands.)

.IP -h
show usage message and exit.

.IP -n
don't execute anything. This option implies -v, but the commands shown
are only an approximation of what would happen, if -n were not given.
This is because the results of reasoning can only be complete, if
commands are executed.

.IP -p
print database of rules and exit. This is not yet really useful
because in fact it exits too early. In the meantime you may want to
try 
.BR "parray ::bras::Rule"
at any place in your script to have a look at all the rules.

.IP -s
silent operation. Don't print anything except warnings and error
messages.

.IP -v
show the full list of commands which are executed.

.IP -ve
installs a verbose version of
.I exec
resulting in a log of all external commands executed. The output is
the closest approximation possible to the typical output of
`make'. However tcl-commands which are executed can only be seen with
-v.


.IP var=value
Your last chance to set environment variables, i.e. to set elements of
Tcl's global array-variable
.IR env .
Typical use of 
.I env
in brasfiles is to use the bras-specific command
.I getenv
to set a (global) variable in way that it can be overridden by the
environment. 

.SH DESCRIPTION
The utility 
.I bras
is written in Tcl and is intended to be used similar to the (in)famous
.I make
utility. A rule file, usually called
.I brasfile
contains rules describing when and how certain commands should be
executed to bring a 
.I target
up-to-date. 
.I Bras
reads the file given with option 
.I  -f
or either
.IR brasfile " or " Brasfile .
Using the rules found in the file,
.I bras 
then determines the commands which are necessary to bring the target(s)
listed on the command line up-to-date. If no target was given, the
target of the first rule found in the file is considered.

Depending on the options used, 
.I bras
executes the derived command list with different levels of verbosity,
shows its reasoning or merely shows what it would do.

.SH SEE ALSO
.nf
|DOC|/bras.tex
|DOC|/bras.ps
|DOC|/bras.pdf
Tcl(n)
.fi


bras/CHANGES
***** 0.0.2 (1997-03-11)
bras.buildCmds: If the target was done already, the `cd $keepPWD' was a bit
early, which resulted in a variable not existing.

The install-script did not enter the path to the installation correctly
into bras.

Simple rule file for SunOS added.

***** 0.0.1 (1997-03-04)
First version with a real version number. Older ones were only identified by
date.

install.wish now allows to specify the tclsh-executable.

Very simple test suite added.

***** old history
o Since tcl7.4, auto_execok seems to have changed its return value. This
  led to a change in evalCmds, proc bras.unknown.

o When exec'ing external commands in bras.unknown, stderr was redirected
  to stdout. This had bad effects when the command itself did redirection. 
----- end of notes for bras.1997-02-25 -----
Enhanced user doc. 
----- end of notes for bras.1997-02-24 -----
A wrong tcl-version in the install script no longer exits the script. It only
results in a warning. 
----- end of notes for bras.1997-02-14 -----
o Corrected bug where a target was not rebuild, if one of its 
  dependencies was rebuild for another target. 
----- end of notes for bras.1997-02-04 -----
A view minor changes to the doc. 
----- end of notes for bras.1997-01-26 -----
New commands Always and include, not yet tested. 
----- end of notes for bras.1996-12-03 -----
Bras.tex continued.
Always-rule implemented. 
----- end of notes for bras.1996-12-01 -----
o Exist-rules now have targets.
o better error reporting
o better -d reporting. 
----- end of notes for bras.1996-11-10 -----
backup pack.
Added GPL to central files.

Started writing doc. Revealed a misconception about Exist-Rules which let
to the decision to also let Exist-Rules have dependencies. Implemented that,
but did not yet test it. The program might even not run through. 
----- end of notes for bras.1996-11-06 -----
backup pack 
----- end of notes for bras.1996-11-05 -----
backup pack 
----- end of notes for bras.1996-11-04 -----
backup pack 
----- end of notes for bras.1996-11-03 -----
backup pack (in statu nascendi) 
----- end of notes for bras.1996-11-02 -----


bras/bras.tcl
########################################################################
#
# This file is part of bras, a program similar to the (in)famous
# `make'-utitlity, written in Tcl.
#
# Copyright (C) 1996-2000 Harald Kirsch
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# $Revision: 1.5 $, $Date: 2000/06/28 19:27:59 $
########################################################################
## source version and package provide
source [file join [file dir [info script]] .version]
#puts "reading [info script]"
##
## This files contains 
## -- the declarations of all namespace-variables
## -- the procs which built up the database,
## -- other misc. procs
##
namespace eval ::bras {

  namespace eval gendep {
    ## This namespace will contain procs which map a target matching a 
    ## pattern rule into a useful dependency. See enterPatternRule for 
    ## more information.
  }

  ##### namespace variables

  ## base
  ##   Directory holding files sourced in by this script, as well as
  ##   files someone might find useful sourcing, in particular those
  ##   with suffix .rules. It is set in .version

  ## Opts
  ##   An array of options controlling bras.
  variable Opts
  set Opts(-d) 0;			# debug output
  set Opts(-k) 0;
  set Opts(-s) 0;			# silent operation
  set Opts(-v) 0;			# verbose operation
  set Opts(-n) 0;			# don't execute any commands
  set Opts(-ve) 0;			# show exec'ed commands

  ## Brasfile
  ##   Name of the rule file used, typically "brasfile" or "Brasfile" 
  variable Brasfile brasfile

  ## Targets
  ##   List of default targets to consider. It is set by the very
  ##   first rule entered into the database

  ## Indent
  ##   Current indent string used by debug messages
  variable Indent ""

  ## Tinfo
  ##   array holding information about targets. With <t> a target,
  ##   <d> its directory, the indices used are:
  ##   <t>,<d>,rule -- index into Rule denoting the rule for <t>,<d>
  ##   <t>,<d>,done -- set, if target was already considered.
  ##                   0: no need to make target
  ##                   1: target will be made
  ##                  -1: target needs to be made, but don't know how

  ## Rule
  ##   database of rules. With integer <i>, the indices are as follows
  ##   <i>,targ -- targets
  ##   <i>,deps -- dependencies
  ##   <i>,cmd  -- command
  ##   all      -- a list of all valid <i>s.
  variable Rule
  set Rule(all) {}

  ## Prule
  ##   database of pattern-rules. Pattern rules are stored and accessed in
  ##   the same order as they are introduced. Therefore each one gets a
  ##   number. The indices are used as follows (<i> denotes an integer):
  ##   all        -- list of all known pattern-rule IDs.
  ##   <i>,trexp  -- regular expression to match target 
  ##   <i>,gendep -- name of dependency generating funciton in
  ##                 ::bras::gendep 
  ##   <i>,cmd    -- command for target/dependency pair
  ##   <i>,cure   -- used by lastMinuteRule, set to 1 if CUrrently
  ##                 REcursively considered.
  variable Prule
  set Prule(all) {}

  ## nextID
  ##   a counter returning unique indices
  variable nextID 0

  ## Known
  ##   array with an element for all files sourced either by following
  ##   an @-target or by an explicit `include'. The directory-part of
  ##   the filename is always normalized by means of [pwd].
  ##

  ## Considering
  ##   is an array with targets as indices. The array element is set
  ##   while ::bras::Consider is working on a target to prevent
  ##   dependency loops

  ## Searchpath
  ## is an array indexed by [pwd] and contains for each directory the
  ## dependency search path. Elements are set by command `searchpath'.

  ## Searched
  ## is an array indexed by [pwd],<name> . If a certain index exists,
  ## <name> is the result of an expansion along brasSearchPath and it
  ## will not be expanded again.

  ## nspace
  ##   internal variable holding the name of the namespace used for
  ##   the next set of predicates and commands. A namespace of this
  ##   name is set up in checkMake, just before the condition of a
  ##   rule is evaluated. All variables a predicate set for the
  ##   command to find later must be set in this namespace.
  variable nspace {}
}

########################################################################
#
# generate a unique number
#
proc ::bras::nextID {} {
  variable nextID
  incr nextID
  return $nextID
}
########################################################################
#
# lappendUnique -- append an element to a list if it is not already in
# there. 
#
proc ::bras::lappendUnique {_list elem} {
  upvar $_list list
  
  if {-1==[lsearch -exact $list $elem]} {
    lappend list $elem
  }
}
########################################################################
#
# concatUnique
#
proc ::bras::concatUnique {_list newElems} {
  upvar $_list list
  
  foreach elem $newElems {
    if {-1!=[lsearch -exact $list $elem]} continue
    lappend list $elem
  }
}
########################################################################
#
# tcl (as of 8.0b1 and previous) does execute a `cd .' thereby
# spoiling its cache for pwd. Since bras happens to execute quite some
# `cd .', calling pwd afterwards, I trick it myself.
#
rename cd _cd
proc cd {dir} {
  if {"$dir"=="."} return
  _cd $dir
}
########################################################################
#
# Depending on command line options, this replaces the normal
# exec-command. Don't be fooled into thinking that this proc is called 
# within namespace ::bras. Option -ve renames it to ::exec.
#
proc ::bras::verboseExec args {
  ::bras::report -ve $args
  return [eval ::bras::exec_orig $args]
}
########################################################################
##
## prints "$text" to either stdout or stderr, depending on
## type. Feel free to redefine this proc, if you want the output to go
## somewhere else.
## The following types are used:
## "warn"  -- reminders for the user about dubious idioms
## "-d", "-v", "-ve" -- output requested by the respective options
## "norm" -- output produces without any option set
##
proc ::bras::report {type text {newline 1} } {
  switch -exact -- $type {
    warn {
      set out stderr
    }
    norm -
    -d -
    -v -
    -ve {
      set out stdout
    }
    default {
      return -code error -errorinfo "wrong type `$type'"
    }
  }
  if {$newline} {
    puts $out $text
  } else {
    puts -nonewline $out $text
  }
}    
########################################################################
##
## gobble
##   a wrapper around `source $file' to handle errors gracefully
##
## In case of error, the line "---SNIP---" is introduced on top of the 
## error stack. Everthing after the first "---SNIP---" is entered onto 
## the stack later by bras-internals and can be deleted before it is
## presented to the user.
##
proc ::bras::gobble {file} {
  global errorInfo errorCode

  if {[catch "uplevel #0 source $file" msg]} {
    ## strip the last 5 lines off the error stack
    set ei [split $errorInfo \n]
    set l [llength $ei]
    set lastline [lindex $ei [expr $l-8]]
    set ei [lrange $ei 0 [expr {$l-9}]]
    regsub "\"\[^\"]+\"" $lastline "\"[file join [pwd] $file]\"" lastline
    lappend ei "${lastline}---SNIP---"
    return -code error -errorinfo [join $ei "\n"]
  }
}
########################################################################
##
## followTarget
##  
## Handle all what is necessary to follow an @-target to its
## home. In particular:
## - change directory
## - read brasfile
## And do all this with the necessary error handling.  
##
## RETURN
##   The current directory (before cd) is returned.
##
proc ::bras::followTarget {target} {
  #puts "followTarget $target"

  set oldpwd [pwd]
  set dir [file dir $target]
  include $dir
  cd [string range $dir 1 end]
  return $oldpwd
}
########################################################################
##
## enterPatternRule
##   Declare a pattern-rule.
##
## PARAMETER
## trexp --
##   regular expression to match a target
## gendep --
##   name of a proc in ::bras::gendep which can map a target which
##   matches trexp into a useful dependency. It must have exactly one
##   paramter which will be the matching target, when it is called.
## bexp --
##   boolean expression to use in a rule derived from this pattern
##   rule 
## cmd --
#    command to attach to a rule derived from this pattern rule
proc ::bras::enterPatternRule {trexp gendep bexp cmd} {
  variable Prule

  ## Emtpy commands are rather useless here
  if {0==[string length $cmd]} {
    return -code error \
	"empty commands are not allowed in pattern rules"
  }

  ## enter the rule
  set id [nextID]
  set Prule(all) [concat $id $Prule(all)]
  set Prule($id,trexp) $trexp
  set Prule($id,gdep)  $gendep
  set Prule($id,bexp)  $bexp
  set Prule($id,cmd)   $cmd
  set Prule($id,cure)  0

  ## create pattern replacement commands for the dependency
  if { 0==[llength [info commands ::bras::gendep::$gendep]] } {
    set body [format {
      set rootname [file rootname $target]
      return [join [list $rootname "%s"] {}]
    } $gendep]
    proc ::bras::gendep::$gendep {target} $body
  }
}
########################################################################
##
## ::bras::enterRule
##    declare a rule
## 
## targets -- list of targets
##    gdep -- a dependency generated in a pattern rule
##    bexp -- a boolean expression 
##     cmd -- a script to generate the target(s)
##
proc ::bras::enterRule {targets gdep bexp {cmd {}} } {
  variable Targets
  variable Rule
  variable Tinfo

  if {[llength $targets]==0} {
    return -code error "The target list may not be empty"
  }

  #puts "enterRule: {$type} {$targets} {$deps} {$cmd} {$bexp}"

  ## if this is the very first explicit rule seen, and if no target was
  ## specified on the command line, this is the default target-list.
  ## It suffices to put just the first element into Targets,
  ## because all of them are made by the command of this rule.
  ## We also record the current directory, because the brasfile may
  ## contain cd-commands.
  if {![info exist Targets]} {
    set Targets [list [pwd] [lindex $targets 0]]
  }

  ## Although more than one `Make'-command for a target is allowed in
  ## a brasfile, all of those are pooled into one rule
  ## internally. Consequently, if a `Make'-command specifies more than
  ## one target which has already a rule associated, they must all
  ## have that same rule.
  set rid {}
  set tmp {}
  set err 0
  foreach t $targets {
    if {[info exist Tinfo($t,[pwd],rule)]} {
      if {"$rid"==""} {
	set rid $Tinfo($t,[pwd],rule)
      } elseif {$rid!=$Tinfo($t,[pwd],rule)} {
	set err 1
      }
      lappend tmp $t
    }
  }
  if {$err} {
    append msg "The targets `$tmp' all have already a rule, but "\
	"these rules are not all same."
    return -code error -errorinfo $msg
  }

  ## If rid is not set now, initialize a rule 
  if {[llength $rid]==0} {
    set rid [nextID]
    lappend Rule(all) $rid
    set Rule($rid,targ) {}
    set Rule($rid,bexp) {}
    set Rule($rid,cmd) {}
  }

  ## We are sure now, all targets either don't have a rule yet or they 
  ## all have the same.
  foreach t $targets {
    set Tinfo($t,[pwd],rule) $rid
  }

  ## Add the new information into Rule($rid,...)
  concatUnique Rule($rid,targ) $targets
  if {"$cmd"!=""} {
    ## It is no good to have more than one command for a target.
    if {""!="$Rule($rid,cmd)" && "$Rule($rid,cmd)"!="$cmd"} {
      set msg {}; append msg \
	  "bras(warning) in `[pwd]': overriding command " \
	  "`$Rule($rid,cmd)' for target `$targets'" \
	  " with `$cmd'"
      report warn $msg
    }
    set Rule($rid,cmd) $cmd

    ## If this rule has a command, we want its boolean expression to
    ## be the first in the list so that it enters its dependencies it
    ## has in front of the dependency list so that [lindex $deps 0] is
    ## equivalent to make's $< .
    set Rule($rid,bexp) \
	[concat [list $targets $gdep $bexp] $Rule($rid,bexp)]
  } else {
    lappend Rule($rid,bexp) $targets $gdep $bexp
  }
  set Rule($rid,run) 0
}
########################################################################
